sources/web/datalab/polymer/components/bread-crumbs/bread-crumbs.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+ /**
+  * Breadcrumb element for Datalab. It takes an array of strings that are
+  * displayed as a horizontal list of links representing a POSIX-like path. A
+  * root link is always displayed at the beginning of the list. Clicking a link
+  * dispatches a 'crumbClicked' custom event that contains the index of the
+  * clicked item. Overflowing links will be hidden from the left side, and an
+  * ellipsis link appears when this is the case. Clicking the ellipsis
+  * dispatches the custom event with the index before the last visible item.
+  * Clicking the root link dispatches a 'rootClicked' custom event.
+  */
+class BreadCrumbsElement extends Polymer.Element {
+
+  public crumbs: string[];
+
+  private lastVisibleIndex: number;
+
+  static get is() { return 'bread-crumbs'; }
+
+  static get properties() {
+    return {
+      crumbs: {
+        observer: '_crumbsChanged',
+        type: Array,
+        value: [],
+      },
+    };
+  }
+
+  ready() {
+    super.ready();
+
+    window.addEventListener('resize', () => this._resizeHandler());
+  }
+
+  _crumbsChanged() {
+    this._resizeHandler();
+  }
+
+  _resizeHandler() {
+    // Let the items render first, before hiding overflowing items
+    Polymer.dom.flush();
+
+    const container = this.$.breadcrumb as HTMLDivElement;
+    const maxWidth = container.offsetWidth - 55;
+
+    const children = container.querySelectorAll('div.part') as NodeListOf<HTMLDivElement>;
+
+    if (!children.length) {
+      return;
+    }
+
+    // Find the index of the last breadcrumb we can fit, starting from the right
+    // We must fit at least the right-most child, so this is the starting point.
+    this.lastVisibleIndex = children.length - 1;
+    children[this.lastVisibleIndex].classList.remove('hidden');
+    let runningWidth = children[this.lastVisibleIndex].offsetWidth;
+
+    for (let i = this.lastVisibleIndex - 1; i >= 0; --i) {
+      if (runningWidth + children[i].offsetWidth < maxWidth) {
```

Comment:
I thought this would make the four loop's definition a little longer than usual, but I'm willing to give it a try if you think it's more readable.

Second solution wouldn't work because we need the loop to terminate after finding the first element that won't fit. Otherwise it'll hide elements in the middle of the path and show earlier elements if they're narrower.